### DISCUSSION TITLE
Create a Shiny interactive qmd document within a Quarto website

### DISCUSSION THREAD

**Discussion by cutterkom** [quarto-dev/quarto-cli#1561] [Q&A] [answered]
I am building a Quarto website (https://quarto.org/docs/websites/) that should have some pages that work interactivly with Shiny.

Assume this mini interactive file which should be named mini-app.qmd:

```
---
title: "Old Faithful"
format: html
server: shiny
---

```{r}
sliderInput("bins", "Number of bins:", 
            min = 1, max = 50, value = 30)
plotOutput("distPlot")
```

```{r}
#| context: server
output$distPlot <- renderPlot({
  x <- faithful[, 2]  # Old Faithful Geyser data
  bins <- seq(min(x), max(x), length.out = input$bins + 1)
  hist(x, breaks = bins, col = 'darkgray', border = 'white')
})
```

This works as a common .qmd file, rendering, preview, everythings great.

But as soon as this file mini-app.qmd is part of a Quarto created with quarto create-project mysite --type website this interactive component isn't working anymore.

The error message:

```
Error in shiny_prerendered_html(input_rmd, render_args) : 
Prerendered HTML file not found at /Users/.../quarto-website/mini-app.html
Calls: .main ... <Anonymous> -> shiny_prerendered_app -> shiny_prerendered_html
Execution halted
```

How can I build a Quarto website that has some pages that are interactive with Shiny? What do I miss?

**Comment by andrewpbray** [comment, marked as answer]
The Quarto website is ultimately a static website - it won't be responding to user input or running any R code. You can embed a shiny app into a qmd that you render in the rstudio IDE and it'll work just fine because the shiny app is running in a session on your computer. Website hosting services like quarto-pub, netlify, and gh-pages will only be serving up the html pages of the site, not running the code necessary to operate the shiny app in response to user input. 

One workaround is to put an iframe into your website that points that page of the website to your shiny app as hosted on a shiny server like shinyapps.io

You can see this behavior in the [quarto website itself](https://quarto.org/docs/interactive/#shiny). The page demonstrating the shiny app is [running it in an iframe](https://github.com/quarto-dev/quarto-web/blob/cca15f68cd229d9413f16847d0445942e533eed3/docs/interactive/index.qmd#L157). You can find that shiny app over on shinyapps.io at <https://gallery.shinyapps.io/goog-trend-index/>.

**Reply by adisarid** [reply]
With WebR and shinylive this answer is no longer true. You can embed a shiny app in a static website and the shiny app will run in the client's browser.

**Reply by G-Accad** [reply]
Can you share some more information/resources on how to do this?

**Reply by adisarid** [reply]
Sure, try [this link](https://github.com/coatless-quarto/r-shinylive-demo).

**Reply by fcamineiro** [reply]
@adisarid, is there an example of this using Shiny for Python (Core) and Quarto? I would like to deploy it on AWS and use Quarto to showcase my portfolio, which is composed of Shiny for Python web apps.

**Reply by adisarid** [reply]
Sure, check out this link for examples:
https://shinylive.io/py/examples/

**Comment by cutterkom** [comment]
Alright, makes sense... and I'll figure something out. Thanks!

**Reply by andrewpbray** [reply]
Sure thing!

Something else I realized: the development of in-browser shiny apps should make possible what you're describing. See <https://shiny.rstudio.com/py/docs/shinylive.html>.

**Reply by eitsupi** [reply]
The python shiny reference page is obviously created in quarto, but comes with an interactive editor and execution environment within the page.
It would be great if there was a way for users to create that kind of documentation on the quarto reference.
https://shiny.rstudio.com/py/docs/overview.html

**Reply by mcanouil** [reply]
The shinylive python tool is [here](https://github.com/rstudio/py-shinylive)

**Comment by mcanouil** [comment]
There is https://shiny.rstudio.com/py/docs/shinylive.html if you want to have a shiny python app working serverless in the browser 🥳
@wch gave a talk today at RStudioConf 2022 about this 🤯

**Comment by jjallaire** [comment]
We'll also be working on a Quarto extension that makes it dead simple to add this to a Quarto page/site.

**Reply by ZellW** [reply]
Can't wait for the extension!

**Reply by mcanouil** [reply]
The extension was released two months ago, see <https://quarto.org/docs/blog/posts/2022-10-25-shinylive-extension/>.

**Reply by aarontovars** [reply]
I am having several problems loading other supported packages (opencv-python, for instance) into shinylive extension in quarto documents. Any idea how to load them?

**Reply by cderv** [reply]
@faaronts probably not the best place to ask. Repo for shinylive extension is here; https://github.com/quarto-ext/shinylive

you can open an issue there to ask about specific shinylive issue. It could also be issue with Pyodide and using `opencv-python` maybe.

Anyway, best to ask in those places than in this discussion IMO
